src/plots/cartesian/make_trace_groups.js

```diff
@@ -0,0 +1,48 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var d3 = require('d3');
+
+/**
+ * helper for cartesian trace types to manage trace groups and plot traces
+ * into them
+ *
+ * @param {div} gd: plot div
+ * @param {object} plotinfo: the cartesian subplot info object
+ * @param {array} cdModule: array of calcdata items for this
+ *     module and subplot combination
+ * @param {d3.selection} traceLayer: a selection containing a single group
+ *     to draw these traces into
+ * @param {string} cls: the class attribute to give each trace group
+ *     so you can give multiple classes separated by spaces
+ * @param {function} plotOneFn: a function that will plot one trace
+ *     takes arguments:
+ *         gd
+ *         plotinfo
+ *         cd: calcdata array for this one trace
+ *         plotGroup: d3 selection of the single group to draw into
+ */
+module.exports = function makeTraceGroups(gd, plotinfo, cdModule, traceLayer, cls, plotOneFn) {
+    var traces = traceLayer.selectAll('g.' + cls.replace(/\s/g, '.'))
+        .data(cdModule, function(cd) { return cd[0].trace.uid; });
+
+    traces.exit().remove();
+
+    traces.enter().append('g')
+        .attr('class', cls);
+
+    traces.each(function(cd) {
```

Comment:
Ah very nice! That `cls.replace(/s/g/, '.')` is a nice touch. 

I think `makeTraceGroups` should work with all svg-based subplot types, so perhaps we could move this to `src/lib/`? No sure where it would belong. There's [`lib/gup.js`](https://github.com/plotly/plotly.js/blob/master/src/lib/gup.js) which has "general-update-pattern" helpers, maybe that would be a good fit.

Personally, I would not have included `plotOneFn` as an argument and here just called `traces.order()`, return `traces` and make `_module.plot` handle the `trace.each` "plot-one" logic. This way we wouldn't be tied down to one `plotOneFn` call signature and we wouldn't have to pass `gd`, `plotinfo` to `makeTraceGroups`, perhaps making things more reusable. But that's really more of a personal opinion, so consider this comment non-blocking :ok_hand:

Reply:
I was trying to match the existing pattern (from `contourPlot.plotWrapper`) but yeah, taking `plotOne` out would be cleaner and even more generalizable, I'll do that. Originally I wasn't returning the final selection but now that we have that this will be easy.